gh-135676: Add a summary of source characters (#138194)
python/cpython: Doc/reference/lexical_analysis.rst

A Python program is read by a *parser*.  Input to the parser is a stream of
:term:`tokens <token>`, generated by the *lexical analyzer* (also known as
the *tokenizer*).
This chapter describes how the lexical analyzer produces these tokens.

.. note::

   A ":dfn:`stream`" is a *sequence*, in the general sense of the word

>    (not necessarily a Python :term:`sequence object <sequence>`).

Reviewer: I'm not sure this note is needed?

Reviewer 2: I agree with @AA-Turner. Stream and sequence are both overloaded terms that may be better unpacked by the reader in context.

Author: OK; I've removed it

The lexical analyzer determines the program text's :ref:`encoding <encodings>`
(UTF-8 by default), and decodes the text into
:ref:`source characters <lexical-source-character>`.

> If the text cannot be decoded, a :exc:`SyntaxError` is raised.

Reviewer 3: Can this also not raise a `Unicode*Error`? For example, running a file with a surrogate:

```
UnicodeEncodeError: 'utf-8' codec can't encode characters in position 1-2: surrogates not allowed
```

Author: This sentence is unchanged from the current docs, and I don't know if it's a docs bug or an implementation bug. Let's leave the issue out of this PR.

> The lexical analyzer then generates a stream of tokens from the source
> characters.
> The type of each generated token, or other special behavior of the analyzer,
> generally depends on the first source character that hasn't yet been processed.
> The following table gives a quick summary of these characters,
> with links to sections that contain more information.

Reviewer 2: ```suggestion
Next, the lexical analyzer uses the source characters to generate a stream of tokens.
The type of a generated token generally depends on the next source character to
be processed. Similarly, other special behavior of the analyzer depends on
the first source character that hasn't yet been processed.
The following table gives a quick summary of these source characters,
with links to sections that contain more information.
```

.. list-table::

>    :header-rows: 1

Reviewer: In general for list tables it can be useful to alternate list markers, e.g. using `-` to denote items of the second-level list. Not essential, though.

Author: All my list-tables will do that from now on :)

   * * Character
     * Next token (or other relevant documentation)

   * * * space
       * tab
       * formfeed
     * * :ref:`Whitespace <whitespace>`

>    * * * CR, LF

Reviewer 3: Should we also list `CRLF`, as we list in the [formal grammar](https://docs.python.org/3.15/reference/lexical_analysis.html#physical-lines):

```
newline: <ASCII LF> | <ASCII CR> <ASCII LF> | <ASCII CR>
```

Author: CRLF *is* these two characters, I don't think we need to repeat them.

     * * :ref:`New line <line-structure>`
       * :ref:`Indentation <indentation>`

   * * * backslash (``\``)
     * * :ref:`Explicit line joining <explicit-joining>`
       * (Also significant in :ref:`string escape sequences <escape-sequences>`)

   * * * hash (``#``)
     * * :ref:`Comment <comments>`

   * * * quote (``'``, ``"``)
     * * :ref:`String literal <strings>`

   * * * ASCII letter (``a``-``z``, ``A``-``Z``)

>        * non-ASCII character

Reviewer: Is 'non-ASCII character' too broad here? Not all characters can form valid identifiers, especially if expanding to the full Unicode space!

Author: It is broad, but: if the tokenizer sees a non-ASCII character, the next token can only be a NAME (or error). (Except inside strings/comments, but then it's not deciding what the next token will be.)

If I remember correctly¹, the tokenizer implementation does lump non-ASCII characters with the letters, and only checks validity *after* it parses an identifier-like token. 

¹ Maybe I don't, but it certainly *could* do that :)

     * * :ref:`Name <identifiers>`
       * Prefixed :ref:`string or bytes literal <strings>`

   * * * underscore (``_``)
     * * :ref:`Name <identifiers>`
       * (Can also be part of :ref:`numeric literals <numbers>`)

   * * * number (``0``-``9``)
     * * :ref:`Numeric literal <numbers>`

   * * * dot (``.``)
     * * :ref:`Numeric literal <numbers>`
       * :ref:`Operator <operators>`

   * * * question mark (``?``)
       * dollar (``$``)
       *
         .. (the following uses zero-width-joiner characters to render
         .. a literal backquote)

>          backquote (``‍`‍``)

Reviewer 4: Could a [substitution](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#substitutions) be used here? See example for nbsp in math.rst.

Author: ReST doesn't support nested markup. I know of no good way to put a substitution in a piece of literal (monospace) text.
